### ISSUE TITLE
Hyperparameter Tuning

### ISSUE THREAD

**Issue by rajithalder-pronto** [roboflow/rf-detr#700]
Hey so i want to like tune hyperparameters for the detection model before a proper training run, is there some package you guys would recommend or ill just use whatever works with a pytorch based framework?

**Comment by isaacrob** [comment]
What parameters are you hoping to tune? Modifying the architecture may interfere negatively with the pretrained weights which would likely decrease performance even if your new parameters are 'better'. For training-based parameters, the defaults should work well. I don't have in mind any particular packages for tuning training parameters, maybe someone else can chime in who uses tools for that.

**Comment by rajithalder-pronto** [comment]
> What parameters are you hoping to tune? Modifying the architecture may interfere negatively with the pretrained weights which would likely decrease performance even if your new parameters are 'better'. For training-based parameters, the defaults should work well. I don't have in mind any particular packages for tuning training parameters, maybe someone else can chime in who uses tools for that.

```python
def objective(trial):
    lr = trial.suggest_float("lr", 1e-5, 5e-4, log=True)
    lr_encoder = trial.suggest_float("lr_encoder", 1e-6, 1e-4, log=True)
    lr_vit_layer_decay = trial.suggest_float("lr_vit_layer_decay", 0.6, 0.95)
    lr_component_decay = trial.suggest_float("lr_component_decay", 0.5, 0.9)
    weight_decay = trial.suggest_float("weight_decay", 1e-5, 1e-3, log=True)
    warmup_epochs = trial.suggest_int("warmup_epochs", 0, 5)
    ema_decay = trial.suggest_float("ema_decay", 0.99, 0.9999, log=True)
    lr_scheduler = trial.suggest_categorical("lr_scheduler", ["step", "cosine"])
    cls_loss_coef = trial.suggest_float("cls_loss_coef", 0.5, 3.0)
    bbox_loss_coef = trial.suggest_float("bbox_loss_coef", 2.0, 8.0)
    giou_loss_coef = trial.suggest_float("giou_loss_coef", 1.0, 4.0)

    output_dir = f"tuning/trial_{trial.number}"
    Path(output_dir).mkdir(parents=True, exist_ok=True)

    model = RFDETRMedium(pretrain_weights=PRETRAIN_WEIGHTS)

    model.train(
        dataset_dir=DATASET_DIR,
        epochs=TUNING_EPOCHS,
        batch_size=2,
        grad_accum_steps=4,
        lr=lr,
        lr_encoder=lr_encoder,
        lr_vit_layer_decay=lr_vit_layer_decay,
        lr_component_decay=lr_component_decay,
        lr_scheduler=lr_scheduler,
        lr_min_factor=0.01,
        warmup_epochs=warmup_epochs,
        weight_decay=weight_decay,
        use_ema=True,
        ema_decay=ema_decay,
        multi_scale=True,
        expanded_scales=True,
        cls_loss_coef=cls_loss_coef,
        bbox_loss_coef=bbox_loss_coef,
        giou_loss_coef=giou_loss_coef,
        checkpoint_interval=TUNING_EPOCHS,
        output_dir=output_dir,
        device="cuda",
    )

    results_path = Path(output_dir) / "results.json"
    with open(results_path) as f:
        results = json.load(f)

    overall = results["class_map"]["valid"][-1]
    map50 = overall["map@50"]
    recall = overall["recall"]
    return map50, recall
```

so i made this function that uses optuna's framework to just go through these hyperparameters updating them each trial to see what fits best.

And im not changing the architecture of any of the models inside,

**Comment by Borda** [comment]
Hey @rajithalder-pronto, thanks for raising this! Hyperparameter tuning — including learning rate scheduling, batch size, and class imbalance handling — is discussed in #296, which has a good thread with practical suggestions.

Closing as a duplicate of #296. Feel free to ask follow-up questions there if your specific tuning scenario isn't covered.
